File: neoforge/src/main/java/com/dreamdisplays/DreamDisplaysMod.java
package com.dreamdisplays;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.dreamdisplays.net.*;
import com.dreamdisplays.render.ScreenWorldRenderer;
import net.minecraft.client.Camera;
import net.minecraft.client.Minecraft;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.commands.CommandSourceStack;
import net.minecraft.network.chat.Component;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.neoforged.bus.api.IEventBus;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.Mod;
import net.neoforged.neoforge.client.event.ClientPlayerNetworkEvent;
import net.neoforged.neoforge.client.event.ClientTickEvent;
import net.neoforged.neoforge.client.event.RegisterClientCommandsEvent;
import net.neoforged.neoforge.client.event.RenderLevelStageEvent;
import net.neoforged.neoforge.common.NeoForge;
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent;
import net.neoforged.neoforge.network.registration.PayloadRegistrar;

@NullMarked
@Mod(value = DreamDisplaysMod.MOD_ID, dist = Dist.CLIENT)
public class DreamDisplaysMod {

    public static final String MOD_ID = "dreamdisplays";

    public DreamDisplaysMod(IEventBus modEventBus) {
        modEventBus.addListener(this::registerPayloads);
        NeoForge.EVENT_BUS.register(this);
    }

    public void registerPayloads(RegisterPayloadHandlersEvent event) {
        PayloadRegistrar registrar = event.registrar(MOD_ID).optional().versioned("1");
        registrar.playBidirectional(DeletePacket.PACKET_ID, DeletePacket.PACKET_CODEC,
            (serverPayload, ctx) -> {
            },
            (clientPayload, ctx) -> PlatformlessInitializer.onDeletePacket(clientPayload)
        );
        registrar.playToClient(DisplayInfoPacket.PACKET_ID, DisplayInfoPacket.PACKET_CODEC,
            (payload, ctx) -> PlatformlessInitializer.onDisplayInfoPacket(payload));

        registrar.playToClient(SyncPacket.PACKET_ID, SyncPacket.PACKET_CODEC,
            (payload, ctx) -> PlatformlessInitializer.onSyncPacket(payload));

        registrar.playToClient(PremiumPacket.PACKET_ID, PremiumPacket.PACKET_CODEC,
            (payload, ctx) -> PlatformlessInitializer.onPremiumPacket(payload));

        registrar.playToServer(RequestSyncPacket.PACKET_ID, RequestSyncPacket.PACKET_CODEC, (p, c) -> {
        });
        registrar.playToServer(ReportPacket.PACKET_ID, ReportPacket.PACKET_CODEC, (p, c) -> {
        });
        registrar.playToServer(VersionPacket.PACKET_ID, VersionPacket.PACKET_CODEC, (p, c) -> {
        });
    }


    @SubscribeEvent
    public void onClientTick(ClientTickEvent.Post event) {
        PlatformlessInitializer.onEndTick(Minecraft.getInstance());
    }

    @SubscribeEvent
    public void onClientStop(ClientPlayerNetworkEvent.LoggingOut event) {
        PlatformlessInitializer.onStop();
    }

    @SubscribeEvent
    public void onRenderLevelAfterEntities(RenderLevelStageEvent.AfterEntities event) {
        Minecraft mc = Minecraft.getInstance();
        if (mc.level == null || mc.player == null) return;

        PoseStack poseStack = event.getPoseStack();
        Camera camera = mc.gameRenderer.getMainCamera();
        ScreenWorldRenderer.render(poseStack, camera);
    }

    @SubscribeEvent
    public void registerCommands(RegisterClientCommandsEvent event) {
        event.getDispatcher().register(
            LiteralArgumentBuilder.<CommandSourceStack>literal("displays")
                .then(LiteralArgumentBuilder.<CommandSourceStack>literal("off")
                    .executes(ctx -> {
                        PlatformlessInitializer.displaysEnabled = false;
                        LocalPlayer p = Minecraft.getInstance().player;
                        if (p != null)
                            p.displayClientMessage(Component.literal("Displays disabled"), false);
                        return 1;
                    }))
                .then(LiteralArgumentBuilder.<CommandSourceStack>literal("on")
                    .executes(ctx -> {
                        PlatformlessInitializer.displaysEnabled = true;
                        LocalPlayer p = Minecraft.getInstance().player;
                        if (p != null) p.displayClientMessage(Component.literal("Displays enabled"), false);
                        return 1;
                    }))
        );
    }

    public void sendPacket(CustomPacketPayload packet) {
        var connection = Minecraft.getInstance().getConnection();
        if (connection != null) {
            connection.send(packet);
        }
    }
}

File: neoforge/src/main/kotlin/com/dreamdisplays/DreamDisplaysMod.kt
package com.dreamdisplays

import com.dreamdisplays.net.c2s.ReportPacket
import com.dreamdisplays.net.c2s.RequestSyncPacket
import com.dreamdisplays.net.common.DeletePacket
import com.dreamdisplays.net.common.DisplayEnabledPacket
import com.dreamdisplays.net.common.SyncPacket
import com.dreamdisplays.net.common.VersionPacket
import com.dreamdisplays.net.s2c.DisplayInfoPacket
import com.dreamdisplays.net.s2c.PremiumPacket
import com.dreamdisplays.net.s2c.ReportEnabledPacket
import com.dreamdisplays.net.s2c.CeilingFloorSupportPacket
import com.dreamdisplays.render.ScreenRenderer
import com.dreamdisplays.screen.managers.ScreenManager
import net.minecraft.client.Minecraft
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.neoforged.api.distmarker.Dist
import net.neoforged.bus.api.IEventBus
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.fml.common.Mod
import net.neoforged.neoforge.client.event.ClientPlayerNetworkEvent
import net.neoforged.neoforge.client.event.ClientTickEvent
import net.neoforged.neoforge.client.event.RenderLevelStageEvent
import net.neoforged.neoforge.common.NeoForge
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent
import org.jspecify.annotations.NullMarked

@NullMarked
@Mod(value = DreamDisplaysMod.MOD_ID, dist = [Dist.CLIENT])
class DreamDisplaysMod(modEventBus: IEventBus) : ModPacketSender {

    init {
        ModInitializer.onModInit(this)
        modEventBus.addListener(this::registerPayloads)
        NeoForge.EVENT_BUS.register(this)
    }

    fun registerPayloads(event: RegisterPayloadHandlersEvent) {
        val registrar = event.registrar(MOD_ID).optional().versioned("1")

        registrar.playBidirectional(
            DeletePacket.PACKET_ID,
            DeletePacket.PACKET_CODEC,
            { _, _ -> },
            { clientPayload, _ -> ModInitializer.onDeletePacket(clientPayload) }
        )

        registrar.playToClient(
            DisplayInfoPacket.PACKET_ID,
            DisplayInfoPacket.PACKET_CODEC
        ) { payload, _ -> ModInitializer.onDisplayInfoPacket(payload) }

        registrar.playToClient(
            PremiumPacket.PACKET_ID,
            PremiumPacket.PACKET_CODEC
        ) { payload, _ -> ModInitializer.onPremiumPacket(payload) }

        registrar.playToClient(
            DisplayEnabledPacket.PACKET_ID,
            DisplayEnabledPacket.PACKET_CODEC
        ) { payload, _ -> ModInitializer.onDisplayEnabledPacket(payload) }

        registrar.playToClient(
            ReportEnabledPacket.PACKET_ID,
            ReportEnabledPacket.PACKET_CODEC
        ) { payload, _ -> ModInitializer.onReportEnabledPacket(payload) }

        registrar.playToClient(
            CeilingFloorSupportPacket.PACKET_ID,
            CeilingFloorSupportPacket.PACKET_CODEC
        ) { payload, _ -> ModInitializer.onCeilingFloorSupportPacket(payload) }

        registrar.playBidirectional(
            SyncPacket.PACKET_ID,
            SyncPacket.PACKET_CODEC,
            { _, _ -> },
            { clientPayload, _ -> ModInitializer.onSyncPacket(clientPayload) }
        )

        registrar.playToServer(
            RequestSyncPacket.PACKET_ID,
            RequestSyncPacket.PACKET_CODEC
        ) { _, _ -> }

        registrar.playToServer(
            ReportPacket.PACKET_ID,
            ReportPacket.PACKET_CODEC
        ) { _, _ -> }

        registrar.playToServer(
            VersionPacket.PACKET_ID,
            VersionPacket.PACKET_CODEC
        ) { _, _ -> }
    }

    @SubscribeEvent
    fun onClientTick(event: ClientTickEvent.Post) {
        ModInitializer.onEndTick(Minecraft.getInstance())
    }

    @SubscribeEvent
    fun onClientStop(event: ClientPlayerNetworkEvent.LoggingOut) {
        ScreenManager.saveAllScreens()
        ScreenManager.unloadAllDisplays()
    }

    @SubscribeEvent
    fun onClientStopping(event: ClientPlayerNetworkEvent.LoggingOut) {
        ModInitializer.onStop()
    }

    @SubscribeEvent
    fun onRenderLevelAfterEntities(event: RenderLevelStageEvent.AfterEntities) {
        val mc = Minecraft.getInstance()
        if (mc.level == null || mc.player == null) return

        val poseStack = event.poseStack
        val camera = mc.gameRenderer.mainCamera
        ScreenRenderer.render(poseStack, camera)
    }

    @SubscribeEvent
    fun onClientLogin(event: ClientPlayerNetworkEvent.LoggingIn) {
        val mc = Minecraft.getInstance()
        if (mc.level != null && mc.player != null) {
            val serverId = if (mc.hasSingleplayerServer()) {
                "singleplayer"
            } else {
                mc.currentServer?.ip ?: "unknown"
            }
            ScreenManager.loadScreensForServer(serverId)
        }
    }

    override fun sendPacket(packet: CustomPacketPayload) {
        val connection = Minecraft.getInstance().connection
        connection?.send(packet)
    }

    companion object {
        const val MOD_ID = "dreamdisplays"
    }
}
